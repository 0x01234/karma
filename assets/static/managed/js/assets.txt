3.2.1-jquery.min.js
2.18.1-moment.min.js
0.3.10-favico.min.js
3.3.7-bootstrap.min.js
3.3.4-bootstrap-switch.min.js
0.2.0-nprogress.min.js
0.8.0-bootstrap-tagsinput.min.js
1.1.1-typeahead.bundle.min.js
4.17.47-bootstrap-datetimepicker.min.js
1.12.2-bootstrap-select.min.js
0.1.2-loaders.css.min.js
2.1.4-js.cookie.min.js
1.8.3-underscore-min.js
4.2.0-masonry.pkgd.min.js
1.6.1-clipboard.min.js
3.15.0-raven.min.js
0.4.1-sha1.min.js
3.0.0-isInViewport.min.js
