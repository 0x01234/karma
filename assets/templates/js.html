<script type="text/javascript" src="{{ .WebPrefix }}static/dist/bundle.js?_={{ .NowQ }}"></script>
<script type="text/javascript" src="{{ .WebPrefix }}static/jquery.typing-0.2.0.js"></script>
