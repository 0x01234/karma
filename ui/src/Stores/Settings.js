import { localStored } from "mobx-stored";

import { action } from "mobx";

const defaultSavedFilters = {
  filters: [],
  present: false
};

class Settings {
  savedFilters = localStored("savedFilters", defaultSavedFilters, {
    delay: 100
  });

  saveFilters = action(newFilters => {
    this.savedFilters.filters = newFilters;
    this.savedFilters.present = true;
  });

  clearSavedFilters = action(() => {
    this.savedFilters.filters = [];
    this.savedFilters.present = false;
  });
}

export { Settings };
