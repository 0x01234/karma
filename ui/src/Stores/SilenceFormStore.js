import { observable, action, computed } from "mobx";

import uniqueId from "lodash.uniqueid";

import moment from "moment";

const NewEmptyMatcher = id => {
  return {
    id: uniqueId(),
    name: "",
    values: [],
    suggestions: {
      names: [],
      values: []
    },
    isRegex: false
  };
};

const ValueToObject = value => ({ label: value, value: value });

class SilenceFormStore {
  // this is used to store modal visibility toggle
  toggle = observable(
    {
      visible: false,
      toggle() {
        this.visible = !this.visible;
      },
      hide() {
        this.visible = false;
      },
      show() {
        this.visible = true;
      }
    },
    { toggle: action.bound, hide: action.bound, show: action.bound }
  );

  // form data is stored here, it's global (rather than attached to the form)
  // so it can be manipulated from other parts of the code
  // example: when user clicks a silence button on alert we should populate
  // this form from that alert so user can easily silence that alert
  data = observable(
    {
      inProgress: false,
      alertmanagers: [],
      matchers: [],
      startsAt: moment(),
      endsAt: moment().add(1, "hour"),
      comment: "",
      author: "",

      resetProgress() {
        this.inProgress = false;
      },

      // append a new empty matcher to the list
      addEmptyMatcher() {
        let m = NewEmptyMatcher();
        this.matchers.push(m);
      },

      deleteMatcher(id) {
        // only delete matchers if we have more than 1
        if (this.matchers.length > 1) {
          this.matchers = this.matchers.filter(m => m.id !== id);
        }
      },

      fillMatchersFromGroup(group) {
        let matchers = [];

        // add matchers for all shared labels in this group
        for (const [key, value] of Object.entries(
          Object.assign({}, group.labels, group.shared.labels)
        )) {
          matchers.push({
            id: uniqueId(),
            name: key,
            values: [ValueToObject(value)],
            suggestions: {
              names: [],
              values: []
            },
            isRegex: false
          });
        }

        // add matchers for all unique labels in this group
        let labels = {};
        for (const alert of group.alerts) {
          for (const [key, value] of Object.entries(alert.labels)) {
            if (!labels[key]) {
              labels[key] = new Set();
            }
            labels[key].add(value);
          }
        }
        for (const [key, values] of Object.entries(labels)) {
          matchers.push({
            id: uniqueId(),
            name: key,
            values: [...values].sort().map(value => ValueToObject(value)),
            suggestions: {
              names: [],
              values: []
            },
            isRegex: values.size > 1
          });
        }

        this.matchers = matchers;
      },

      verifyStarEnd() {
        if (this.endsAt.isSameOrBefore(this.startsAt)) {
          this.endsAt = moment(this.startsAt).add(1, "minutes");
        }
      },
      incStart(minutes) {
        this.startsAt = moment(this.startsAt).add(minutes, "minutes");
        this.verifyStarEnd();
      },
      decStart(minutes) {
        this.startsAt = moment(this.startsAt).subtract(minutes, "minutes");
        this.verifyStarEnd();
      },

      incEnd(minutes) {
        this.endsAt = moment(this.endsAt).add(minutes, "minutes");
        this.verifyStarEnd();
      },
      decEnd(minutes) {
        this.endsAt = moment(this.endsAt).subtract(minutes, "minutes");
        this.verifyStarEnd();
      },

      incDuration(minutes) {
        this.endsAt = moment(this.endsAt).add(minutes, "minutes");
      },
      decDuration(minutes) {
        const newEndsAt = moment(this.endsAt).subtract(minutes, "minutes");
        if (newEndsAt.isAfter(this.startsAt)) {
          this.endsAt = newEndsAt;
        }
      },

      get toAlertmanagerPayload() {
        const payload = {
          matchers: this.matchers.map(m => ({
            name: m.name,
            value:
              m.values.length > 1
                ? `(${m.values.map(v => v.value).join("|")})`
                : m.values.length === 1
                  ? m.values[0].value
                  : "",
            isRegex: m.isRegex
          })),
          startsAt: this.startsAt
            .second(0)
            .millisecond(0)
            .toISOString(),
          endsAt: this.endsAt
            .second(59)
            .millisecond(0)
            .toISOString(),
          createdBy: this.author,
          comment: this.comment
        };
        return payload;
      },
      get toDuration() {
        const data = {
          days: this.endsAt.diff(this.startsAt, "days"),
          hours: this.endsAt.diff(this.startsAt, "hours") % 24,
          minutes: this.endsAt.diff(this.startsAt, "minutes") % 60
        };
        return data;
      }
    },
    {
      resetProgress: action.bound,
      addEmptyMatcher: action.bound,
      deleteMatcher: action.bound,
      fillMatchersFromGroup: action.bound,
      verifyStarEnd: action.bound,
      incStart: action.bound,
      decStart: action.bound,
      incEnd: action.bound,
      decEnd: action.bound,
      incDuration: action.bound,
      decDuration: action.bound,
      toAlertmanagerPayload: computed,
      toDuration: computed
    },
    { name: "Silence form store" }
  );
}

export { SilenceFormStore };
