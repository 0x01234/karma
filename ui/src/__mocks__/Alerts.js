const MockAnnotation = (name, value, visible, isLink) => ({
  name: name,
  value: value,
  visible: visible,
  isLink: isLink
});

const MockAlert = (annotations, labels, state) => ({
  annotations: annotations,
  labels: labels,
  startsAt: "2018-08-14T17:36:40.017867056Z",
  endsAt: "0001-01-01T00:00:00Z",
  state: state,
  alertmanager: [
    {
      name: "default",
      state: "active",
      startsAt: "2018-08-14T17:36:40.017867056Z",
      endsAt: "0001-01-01T00:00:00Z",
      source: "localhost/prometheus",
      silencedBy: []
    }
  ],
  receiver: "by-name"
});

const MockAlertGroup = (
  rootLabels,
  alerts,
  sharedAnnotations,
  sharedLabels
) => ({
  receiver: "by-name",
  labels: rootLabels,
  alerts: alerts,
  id: "099c5ca6d1c92f615b13056b935d0c8dee70f18c",
  hash: "53a4bb3d7e916450b3bda550976f9578db5b2ad3",
  alertmanagerCount: {
    default: 1
  },
  stateCount: {
    active: 1,
    suppressed: 0,
    unprocessed: 0
  },
  shared: {
    annotations: sharedAnnotations,
    labels: sharedLabels
  }
});

export { MockAnnotation, MockAlert, MockAlertGroup };
