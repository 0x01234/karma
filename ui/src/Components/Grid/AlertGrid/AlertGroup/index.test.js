import React from "react";

import { Provider } from "mobx-react";

import { mount } from "enzyme";

import moment from "moment";

import { MockAlert, MockAlertGroup } from "__mocks__/Alerts.js";
import { AlertStore } from "Stores/AlertStore";
import { Settings } from "Stores/Settings";
import { SilenceFormStore } from "Stores/SilenceFormStore";
import { AlertGroup } from ".";

let alertStore;
let settingsStore;
let silenceFormStore;
let group;

const MockGroup = groupName => {
  const group = MockAlertGroup(
    { alertname: "Fake Alert", groupName: groupName },
    [],
    [],
    {},
    {}
  );
  return group;
};

beforeEach(() => {
  alertStore = new AlertStore([]);
  settingsStore = new Settings();
  silenceFormStore = new SilenceFormStore();
  group = MockGroup("fakeGroup");
});

const MockAlerts = alertCount => {
  for (let i = 1; i <= alertCount; i++) {
    let alert = MockAlert([], { instance: `instance${i}` }, "active");
    const startsAt = moment().toISOString();
    alert.startsAt = startsAt;
    alert.alertmanager[0].startsAt = startsAt;
    group.alerts.push(alert);
  }
};

const MountedAlertGroup = (afterUpdate, showAlertmanagers) => {
  return mount(
    <Provider alertStore={alertStore}>
      <AlertGroup
        afterUpdate={afterUpdate}
        group={group}
        showAlertmanagers={showAlertmanagers}
        settingsStore={settingsStore}
        silenceFormStore={silenceFormStore}
      />
    </Provider>
  );
};

describe("<AlertGroup />", () => {
  it("renders Alertmanager labels in footer if showAlertmanagersInFooter=true", () => {
    MockAlerts(2);
    const tree = MountedAlertGroup(jest.fn(), true).find("AlertGroup");
    expect(tree.find("GroupFooter").html()).toMatch(/@alertmanager/);
  });

  it("doesn't render alertmanager labels in footer when they are unique", () => {
    MockAlerts(5);
    for (let i = 0; i < group.alerts.length; i++) {
      group.alerts[i].alertmanager[0].name = `fakeAlertmanager${i}`;
    }
    group.alertmanagerCount = {
      fakeAlertmanager0: 1,
      fakeAlertmanager1: 1,
      fakeAlertmanager2: 1,
      fakeAlertmanager3: 1,
      fakeAlertmanager4: 1
    };
    const tree = MountedAlertGroup(jest.fn(), true);

    const alerts = tree.find("ul.list-group");
    expect(alerts.html()).toMatch(/@alertmanager/);

    const footer = tree.find("GroupFooter");
    expect(footer.html()).not.toMatch(/@alertmanager/);
  });

  it("only renders titlebar when collapsed", () => {
    MockAlerts(10);
    const tree = MountedAlertGroup(jest.fn(), false);
    const alertGroup = tree.find("AlertGroup");
    alertGroup.instance().collapse.toggle();
    expect(alertGroup.instance().collapse.value).toBe(true);
    tree.update();
    expect(tree.find("Alert")).toHaveLength(0);
    expect(tree.find("ul.list-group")).toHaveLength(0);
  });
});

const ValidateLoadButtonPresent = (totalAlerts, isPresent) => {
  MockAlerts(totalAlerts);
  const tree = MountedAlertGroup(jest.fn(), false);
  const buttons = tree.find("button");
  expect(buttons).toHaveLength(isPresent ? 2 : 0);
};

const ValidateLoadButtonAction = (
  totalAlerts,
  buttonIndex,
  iconMatch,
  loadedAlerts,
  alertsToRenderBeforeClick
) => {
  MockAlerts(totalAlerts);
  const tree = MountedAlertGroup(jest.fn(), false);
  if (alertsToRenderBeforeClick !== undefined) {
    tree
      .find("AlertGroup")
      .instance().renderConfig.alertsToRender = alertsToRenderBeforeClick;
    tree.update();
  }
  const loadMore = tree.find("button").at(buttonIndex);
  expect(loadMore.html()).toMatch(iconMatch);
  loadMore.simulate("click");
  tree.update();
  expect(tree.find("Alert")).toHaveLength(loadedAlerts);
};

describe("<AlertGroup /> renderConfig", () => {
  it("settingsStore.alertGroupConfig.config.defaultRenderCount should be 5 by default", () => {
    expect(settingsStore.alertGroupConfig.config.defaultRenderCount).toBe(5);
  });

  it("renderConfig.alertsToRender should be 5 by default", () => {
    const tree = MountedAlertGroup(jest.fn(), false).find("AlertGroup");
    expect(tree.instance().renderConfig.alertsToRender).toBe(5);
  });

  it("renders only up to renderConfig.alertsToRender alerts", () => {
    MockAlerts(50);
    const tree = MountedAlertGroup(jest.fn(), false).find("AlertGroup");
    const alerts = tree.find("Alert");
    expect(alerts).toHaveLength(tree.instance().renderConfig.alertsToRender);
  });

  it("load buttons are not rendered for 1 alert", () => {
    ValidateLoadButtonPresent(1, false);
  });

  it("load buttons are not rendered for 5 alerts", () => {
    ValidateLoadButtonPresent(5, false);
  });

  it("load buttons are rendered for 6 alert", () => {
    ValidateLoadButtonPresent(6, true);
  });

  it("clicking - icon hides 1 alert if there's 6 in total", () => {
    ValidateLoadButtonAction(6, 0, /fa-minus/, 5, 6);
  });

  it("clicking - icon hides 1 alert if there's 6 in total and we're showing 3", () => {
    ValidateLoadButtonAction(6, 0, /fa-minus/, 2, 3);
  });

  it("clicking - icon hides 2 alerts if there's 7 in total and we're showing 7", () => {
    ValidateLoadButtonAction(7, 0, /fa-minus/, 5, 7);
  });

  it("clicking - icon hides 5 alerts if there's 10 in total and we're showing 10", () => {
    ValidateLoadButtonAction(10, 0, /fa-minus/, 5, 10);
  });

  it("clicking - icon hides 5 alerts if there's 18 in total and we're showing 17", () => {
    ValidateLoadButtonAction(18, 0, /fa-minus/, 12, 17);
  });

  it("clicking + icon loads 1 more alert if there's 6 in total", () => {
    ValidateLoadButtonAction(6, 1, /fa-plus/, 6);
  });

  it("clicking + icon loads 4 more alert if there's 9 in total", () => {
    ValidateLoadButtonAction(9, 1, /fa-plus/, 9);
  });

  it("clicking + icon loads 5 more alert if there's 14 in total", () => {
    ValidateLoadButtonAction(14, 1, /fa-plus/, 10);
  });

  it("clicking + icon loads 5 more alert if there's 25 in total and we're showing 16", () => {
    ValidateLoadButtonAction(25, 1, /fa-plus/, 22, 17);
  });
});
