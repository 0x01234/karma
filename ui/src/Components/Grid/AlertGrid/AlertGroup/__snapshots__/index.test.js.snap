// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AlertGroupContent /> lazy rendering matches snapshot when invisible 1`] = `
"
<div class=\\"components-grid-alertgrid-alertgroup p-1 components-animation-fade-appear components-animation-fade-appear-active\\">
  <div class=\\"card bg-light\\">
    <h5 class=\\"card-header mb-0 d-flex flex-row px-2 py-1 \\">
      <span class=\\"flex-shrink-0 flex-grow-0\\">
        <span class=\\"text-muted cursor-pointer badge pl-0 components-label mr-0 components-grid-alertgroup-099c5ca6d1c92f615b13056b935d0c8dee70f18c\\"
              data-toggle=\\"dropdown\\"
        >
          <svg aria-hidden=\\"true\\"
               focusable=\\"false\\"
               data-prefix=\\"fas\\"
               data-icon=\\"ellipsis-v\\"
               class=\\"svg-inline--fa fa-ellipsis-v fa-w-6 \\"
               role=\\"img\\"
               xmlns=\\"http://www.w3.org/2000/svg\\"
               viewbox=\\"0 0 192 512\\"
          >
            <path fill=\\"currentColor\\"
                  d=\\"M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z\\"
            >
            </path>
          </svg>
        </span>
      </span>
      <span class=\\"flex-shrink-1 flex-grow-1\\"
            style=\\"min-width: 0;\\"
      >
        <div class
             style=\\"display: inline-block; max-width: 100%;\\"
             data-tooltipped
             aria-describedby=\\"tippy-tooltip-345\\"
             data-original-title=\\"Click to only show alerts with this label or Alt+Click to hide them\\"
        >
          <span class=\\"components-label badge badge-dark components-label-dark components-label-with-hover\\">
            <span class=\\"components-label-name\\">
              alertname:
            </span>
            <span class=\\"components-label-value\\">
              Fake Alert
            </span>
          </span>
        </div>
        <div class
             style=\\"display: inline-block; max-width: 100%;\\"
             data-tooltipped
             aria-describedby=\\"tippy-tooltip-346\\"
             data-original-title=\\"Click to only show alerts with this label or Alt+Click to hide them\\"
        >
          <span class=\\"components-label badge badge-warning components-label-dark components-label-with-hover\\">
            <span class=\\"components-label-name\\">
              groupName:
            </span>
            <span class=\\"components-label-value\\">
              fakeGroup
            </span>
          </span>
        </div>
      </span>
      <span class=\\"flex-shrink-0 flex-grow-0 ml-auto pl-1\\">
        <div class
             style=\\"display: inline-block; max-width: 100%;\\"
             data-tooltipped
             aria-describedby=\\"tippy-tooltip-347\\"
             data-original-title=\\"Click to only show active alerts or Alt+Click to hide them\\"
        >
          <span class=\\"components-label badge badge-danger badge-pill components-label-with-hover\\">
            1
          </span>
        </div>
        <span class=\\"text-muted cursor-pointer badge px-0 components-label mr-0\\">
          <div class
               style=\\"display: inline-block; max-width: 100%;\\"
               data-tooltipped
               aria-describedby=\\"tippy-tooltip-348\\"
               data-original-title=\\"Toggle group details\\"
          >
            <svg aria-hidden=\\"true\\"
                 focusable=\\"false\\"
                 data-prefix=\\"fas\\"
                 data-icon=\\"chevron-down\\"
                 class=\\"svg-inline--fa fa-chevron-down fa-w-14 \\"
                 role=\\"img\\"
                 xmlns=\\"http://www.w3.org/2000/svg\\"
                 viewbox=\\"0 0 448 512\\"
            >
              <path fill=\\"currentColor\\"
                    d=\\"M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z\\"
              >
              </path>
            </svg>
          </div>
        </span>
      </span>
    </h5>
    <div class=\\"card-body bg-white px-2 py-1\\">
      <ul class=\\"list-group\\">
        <li class=\\"components-grid-alertgrid-alertgroup-alert list-group-item pl-1 pr-0 py-0 my-1 rounded-0 border-left-1 border-right-0 border-top-0 border-bottom-0 border-light\\">
          <div class=\\"components-label badge badge-light text-light\\"
               style=\\"width: 100%; height: 1.3rem;\\"
          >
          </div>
          <div class=\\"components-label badge badge-secondary text-secondary\\"
               style=\\"width: 70px; height: 1.3rem;\\"
          >
          </div>
          <div class=\\"components-label badge badge-light text-light\\"
               style=\\"width: 150px; height: 1.3rem;\\"
          >
          </div>
          <div class=\\"components-label badge badge-light text-light\\"
               style=\\"width: 80px; height: 1.3rem;\\"
          >
          </div>
          <div class=\\"components-label badge badge-light text-light\\"
               style=\\"width: 70px; height: 1.3rem;\\"
          >
          </div>
        </li>
        <li class=\\"components-grid-alertgrid-alertgroup-alert list-group-item pl-1 pr-0 py-0 my-1 rounded-0 border-left-1 border-right-0 border-top-0 border-bottom-0 border-light\\">
          <div class=\\"components-label badge badge-light text-light\\"
               style=\\"width: 100%; height: 1.3rem;\\"
          >
          </div>
          <div class=\\"components-label badge badge-secondary text-secondary\\"
               style=\\"width: 70px; height: 1.3rem;\\"
          >
          </div>
          <div class=\\"components-label badge badge-light text-light\\"
               style=\\"width: 80px; height: 1.3rem;\\"
          >
          </div>
          <div class=\\"components-label badge badge-light text-light\\"
               style=\\"width: 90px; height: 1.3rem;\\"
          >
          </div>
          <div class=\\"components-label badge badge-light text-light\\"
               style=\\"width: 120px; height: 1.3rem;\\"
          >
          </div>
        </li>
      </ul>
    </div>
  </div>
</div>
"
`;
